### ISSUE TITLE
PowerShell deployment example

### ISSUE THREAD

**Issue by austoonz** [awslabs/aws-lambda-powershell-runtime#16]
Provide an example for how to deploy the runtime using PowerShell (and only PowerShell).

**Comment by Conrad-awaze** [comment]
Is this a request to show how to deploy and use the runtime? This is what I'm after too so if there are any instructions I'd me most grateful.

**Comment by gw-bcharboneau** [comment]
> Is this a request to show how to deploy and use the runtime? This is what I'm after too so if there are any instructions I'd me most grateful.

yes it is. It is to offer an example of how to deploy the custom runtime, using nothing but powershell cmdlets/modules.

I am patiently awaiting this.

**Comment by austoonz** [comment]
Every time we try to go down this route for a pure Windows PowerShell build, the underlying Lambda layer is not formatted correctly. For example the `/opt/bootstrap` file won't be marked with the executable bit in Linux. Or the CloudFormation template ends up UTF8 with BOM, even though [we are forcing](https://github.com/awslabs/aws-lambda-powershell-runtime/blob/main/examples/demo-runtime-layer-function/build-layers.ps1#L27) `utf8` with the `Set-Content` cmdlet.

The only reliable way we've currently found to build from Windows is to use WSL from Windows, and the SAM CLI from within WSL.

**Comment by mtboren** [comment]
> ...reliable way we've currently found to build from Windows...

Greets. Firstly, still loving this Lambda PS runtime -- so good!

As for building/deploying a working runtime layer from Windows directly (as in, _not_ from within WSL), I've been able to succeed in at least a couple of ways. In effort to see how I can help with this issue, I'll share those here to get feedback.

❓ My question to the group: what, beyond the descriptions below, will be most helpful in enabling people to succeed in this use case of, "new Lambda PS runtime layer built and deployed from 'directly' within Windows"? Like, make a repo with steps/examples of end-to-end success? Or, just a PR to this current repo with a Docs update? Something else?

Success from Windows via each of:
- perform all of the operations (listed below) from app "PowerShell" (v7.x) within Windows (I'm wondering if some failures for people have been due to legacy / PSv5 behaviors)
- perform all of the same operations from app "Windows PowerShell" (v5.1) within Windows, with one modification of the [/build-PwshRuntimeLayer.ps1:76](https://github.com/awslabs/aws-lambda-powershell-runtime/blob/main/powershell-runtime/build-PwshRuntimeLayer.ps1#L76) file:
    - change the `-Encoding` param value to `Ascii` from `UTF8` (prevents the aforementioned BOM issue)

Steps:
- `git clone https://github.com/awslabs/aws-lambda-powershell-runtime`
- build local layer contents via `powershell-runtime/build-PwshRuntimeLayer.ps1`
- deploy with SAM via `sam deploy --template powershell-runtime/template.yml --<rest of "profile" and "s3bucket" params>`
- deploy a Lambda function stack with the new runtime layer as its layer
- invoke Lambda function
- profit 💰

**Comment by austoonz** [comment]
Thank you for the details.

I will note that in both a Windows Workspace and a Windows EC2 Instance of mine, the above does not work from either PowerShell or Windows PowerShell (after changing the `-Encoding` parameter). Both end up failing with the following error, which is indicating the execute bit of the bootstrap file is not set correctly.

For me, using the PowerShell build script in PowerShell on Linux works as expected. This makes me think that it has something to do with the `sam deploy` command on Windows causing the execute bit on the bootstrap file to be dropped. I'll dig in further on that...

```json
{
  "errorType": "Runtime.InvalidEntrypoint",
  "errorMessage": "RequestId: 665609a7-17d4-496c-b052-1b0c463561b2 Error: fork/exec /opt/bootstrap: exec format error"
}
```

**Comment by austoonz** [comment]
For awareness from further testing, I think this error might have something to do with the `pwsh` binary, not the bootstrap file.

In my Linux testing, after building the layer with `sam build`, I can run `./.aws-sam/build/PwshRuntimeLayer/powershell/pwsh -version` with success. However in the build using PowerShell, `./layers/runtimeLayer/powershell/pwsh -version` results in `layers/runtimeLayer/powershell/pwsh: cannot execute binary file`.

That's where I'm now looking...

**Comment by austoonz** [comment]
Found what I think the issue is, will test further across Windows, Linux, and PowerShell/Windows PowerShell to confirm. The `examples/demo-runtime-layer-function` which I'm testing with is telling the [PowerShell runtime build script](https://github.com/awslabs/aws-lambda-powershell-runtime/blob/main/examples/demo-runtime-layer-function/build-layers.ps1#L12) to build using `arm64`, instead of `x64`.

**Comment by mtboren** [comment]
> ...have complete PowerShell examples that do not depend on `sam deploy`...

Sounds good! @austoonz -- for the packaging portion of such an example, OK to use `sam package` for the <zip up local resources, upload to S3, and update SAM template with resultant `ContentUri:` value>? Or are you picturing an example that uses a strictly CloudFormation template (no `Transform:` property, no SAM whatsoever)?

**Comment by mtboren** [comment]
> ...an example that uses a strictly CloudFormation template (no `Transform:` property, no SAM whatsoever..

In case this is the desire, I forked this repo and pushed a couple of commits that add such an example. See the comparison at https://github.com/awslabs/aws-lambda-powershell-runtime/compare/main..mtboren:aws-lambda-powershell-runtime_fork:feat_AddPShellDeployExample

How does that do for including a PowerShell-only deployment example, @austoonz?
